### DISCUSSION TITLE
Triggering by other repository

### DISCUSSION THREAD

**Discussion by daid** [community/community#26323] [Actions]
<p>Is it possible to run a github action on a push to a 2nd repository?</p>
<p>For example, I have 2 repos:</p>
<ul>
<li>Application</li>
<li>Library</li>
</ul>
<p>As soon as I push to Library, I would like the Library CI to run, but also run the CI for the Application, to see if the application still works with the changes in the Library.</p>

**Comment by ethomson** [comment, marked as answer]
<p>Triggers are per-repository, _however _you could <a href="https://help.github.com/en/articles/events-that-trigger-workflows" rel="nofollow noopener">send a “repository dispatch” event to your Application repository</a>.  This is a custom event type that has no explicit meaning within GitHub, that you can use for your own purposes.</p>
<p>So in the application’s workflow, you could listen for a repository dispatch event:</p>
<pre><code>on: repository_dispatch
</code></pre>
<p>And then in the library’s workflow, you could send a repository dispatch event.  You’ll need to set up a personal access token as a secret in the library’s repository:</p>
<pre><code>run: |
curl -XPOST -u "${{ secrets.PAT_USERNAME}}:${{secrets.PAT_TOKEN}}" -H "Accept: application/vnd.github.everest-preview+json" -H "Content-Type: application/json" https://api.github.com/repos/YOURNAME/APPLICATION_NAME/dispatches --data '{"event_type": "build_application"}'
</code></pre>
<p>Replace YOURNAME/APPLICATION_NAME with the name of your github repository for your Application.</p>
<p>Set up a secret named PAT_USERNAME with your username, and PAT_TOKEN with the personal access token that you created (it needs repo level access).</p>

**Reply by couling** [reply]
It's such a shame this is _"the wrong way around"_. The required trigger is usually re-build the dependant when the dependency just completed. As a general rule, dependants know about their own dependencies.

**Reply by alelom** [reply]
Answer with up-to-date POST: https://github.com/orgs/community/discussions/26323#discussioncomment-3251448

**Reply by adrian-85** [reply]
I realize this is a few years old, but wondering if GitHub has come up with a more specific method since as their documentation seems to only suggest this event type for non-GitHub repos?

```.md
You can use the GitHub API to trigger a webhook event called [repository_dispatch](https://docs.github.com/en/webhooks-and-events/webhooks/webhook-events-and-payloads#repository_dispatch) when you want to trigger a workflow for activity that happens outside of GitHub.
```
https://docs.github.com/en/actions/writing-workflows/choosing-when-your-workflow-runs/events-that-trigger-workflows#repository_dispatch

**Reply by lr6620573-png** [reply]
> Triggers are per-repository, _however _you could [send a “repository dispatch” event to your Application repository](https://help.github.com/en/articles/events-that-trigger-workflows).  This is a custom event type that has no explicit meaning within GitHub, that you can use for your own purposes.
> 
> So in the application’s workflow, you could listen for a repository dispatch event:
> 
> ```
> on: repository_dispatch
> ```
> 
> And then in the library’s workflow, you could send a repository dispatch event.  You’ll need to set up a personal access token as a secret in the library’s repository:
> 
> ```
> run: |
> curl -XPOST -u "${{ secrets.PAT_USERNAME}}:${{secrets.PAT_TOKEN}}" -H "Accept: application/vnd.github.everest-preview+json" -H "Content-Type: application/json" https://api.github.com/repos/YOURNAME/APPLICATION_NAME/dispatches --data '{"event_type": "build_application"}'
> ```
> 
> Replace YOURNAME/APPLICATION_NAME with the name of your github repository for your Application.
> 
> Set up a secret named PAT_USERNAME with your username, and PAT_TOKEN with the personal access token that you created (it needs repo level access).

seriusly i should diet better offline aset

**Comment by daid** [comment]
<p>Thanks. I’ve accepted it as a sollution.</p>
<p>Even if it doesn’t really work for my specific usecase, it does seem like a valid way to solve my question.</p>
<p>I have 1 library that is used by about 10 different repositories. And I don’t want to add knowledge about those external repositories to my main library repository. My home grown CI solution currently allows this “watch link”, but I understand that it’s a bit of an odd usecase.</p>

**Comment by stefan2904** [comment]
<p>I don’t agree that this is a “odd” usecase.</p>
<p>Actually it is very common that builds in upstream projects (e.g. libraries) should also trigger a build. Especially true in the CD world.</p>

**Reply by mahiki** [reply]
in my case its a package, and a `homebrew-tap` repo that I want to update with the latest relase.

**Reply by lr6620573-png** [reply]
i can only prefer to actully express the homebrew rather than drink it and its a lemonade stand at first soury sweet ride though wise to offer it out for a ring at least i meant someone to lassie once but cauche a bra where a martry and the pirates at

**Comment by yonicd** [comment]
<p>Is it possible to go the other way.</p>
<p>If my repository depends on 4 other repos can an action be triggered if one of them has a commit action (to the master for example.)?</p>

**Reply by lr6620573-png** [reply]
sure go on go ...

**Comment by lwaldron** [comment]
<p>This POST command seems to be out of date - see <a href="https://docs.github.com/en/rest/reference/actions#create-a-workflow-dispatch-event" rel="nofollow noopener">https://docs.github.com/en/rest/reference/actions#create-a-workflow-dispatch-event</a>. I got it working instead with this in the application workflow:</p>
<pre><code class="lang-auto">on: workflow_dispatch
</code></pre>
<p>and this curl command in the library (no more “event_type”):</p>
<pre><code class="lang-auto">curl -XPOST -u "${{ secrets.PAT_USERNAME}}:${{secrets.PAT_TOKEN}}" -H "Accept: application/vnd.github.everest-preview+json" -H "Content-Type: application/json" https://api.github.com/repos/YOURNAME/APPLICATION_NAME/actions/workflows/build.yaml/dispatches --data '{"ref": "master"}'
</code></pre>
<p>I also found that the only permission required for the token was “public_repo Access public repositories”. Lot of trial and error so hopefully this will save someone else the time.</p>

**Reply by ndrewtl** [reply]
Do you by any chance if this works for private repos? I'm not sure how to authorize a dispatch when the repository is private

**Reply by iglesiasroberto14** [reply]
@ndrewtl Did you find out if it works for private repos?

**Reply by rmoff** [reply]
I got this working—write-up here if you're interested: https://rmoff.net/2024/01/17/antora-deployment-to-cloudflare-across-private-repositories-with-github-actions/

**Reply by DerekChia** [reply]
yes it works with private repo

**Reply by lizozom** [reply]
I used the version from the github docs (notification from a public repo to a private one, no user name required)
https://docs.github.com/en/rest/actions/workflows?apiVersion=2022-11-28#create-a-workflow-dispatch-event

**Comment by sdaly2107** [comment]
<p>This is pretty important for our CI across an SDK which is split into various modules across different repos.  It’s something that’s supported in Jenkins.  Hopefully this will be functionality that arrives at some point.</p>
<p>We also don’t want to hardcode details about consumer modules in our library modules.  One solution could be to use a webhook, then from the webhook invoke the workflows using repo dispatch.  Feels like quite a bit of effort to set this up though and dependency context would need sorted somewhere else too.</p>
<p>Another option could be to have an intermediate repo which holds workflows which hold teh dependency contexts, so the lib workflow simply uses repository dispatch on a workflow in the intermediate repo.  That workflow then does a repo dispatch on the required workflows in other repos.  If dependencies change, then they can be updated in one place.</p>

**Comment by Ayfri** [comment]
<p>How can I use it for two repositories on an organization?</p>

**Comment by Mause** [comment]
<p>You can do something like this nowadays:</p>
<pre><code class="lang-yaml">    - name: Deploy Stage
      uses: actions/github-script@v5
      with:
        script: |
          await github.rest.actions.createWorkflowDispatch({
            'Mause',
            'fdroidrepo',
            'workflows/update.yaml',
            'main',
          });
</code></pre>

**Reply by iosifache** [reply]
Maybe this would help someone. Here you can find the technique applied: [producer workflow](https://github.com/Jax-Core/ReadME-Template/blob/629028c10fae8581afd0547c31364dc1fcf3e8ec/.github/workflows/Update-CoreRepos.yml) and [consumer workflow](https://github.com/Jax-Core/JaxCore/blob/main/.github/workflows/Dynamic-Readme.yml).

**Reply by phun-ky** [reply]
@iosifache we're getting closer, but from what I can read of the code, the "producer workflow" will trigger on _any_ `workflow_dispatch`? How can we identify and conditionally trigger another action? As in: action `build` on repository A is complete, it should only trigger action `build` on repository B

**Reply by iosifache** [reply]
@phun-ky, yes, in that exact situation. It will be triggered on `workflow_dispatch`, which is the event that occurs when the action is manually triggered.

I don't know what you mean by "identifying", but you can conditionally trigger the action by (1) modifying the `on` property of the "producer" action, or (2) using the `steps` of the action to aboard its execution (such that `github.rest.actions.createWorkflowDispatch` is not called anymore).

I'm not sure what you mean by "identifying," but you may conditionally start the action by (1) altering the "producer" action's `on` property, or (2) utilizing the action's `steps` to control its execution (such that `github.rest.actions.createWorkflowDispatch` is no longer invoked).

P.S.: I've provided a link to the actions I was working on one year ago. Perhaps they will be useful to you: [the "producer" action](https://github.com/MutableSecurity/mutablesecurity/blob/main/.github/workflows/website_trigger.yaml) and [the "consumer" action](https://github.com/MutableSecurity/website/blob/main/.github/workflows/update_autodocs.yml).

**Comment by jumarko** [comment]
<p>As far as I understand, this still uses the default repo-scoped GITHUB_TOKEN so you need to pass your own token if you want to access a different repository: <a href="https://github.com/actions/github-script#using-a-separate-github-token" class="inline-onebox" rel="noopener nofollow ugc">GitHub - actions/github-script: Write workflows scripting the GitHub API in JavaScript</a></p>

**Comment by CartBlanche** [comment]
<p>Also worth noting that the documentation link to the right area is <a href="https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#repository_dispatch" rel="noopener nofollow ugc">https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#repository_dispatch</a></p>

**Comment by up9cloud** [comment]
<p>A complete use case</p>
<pre><code class="lang-yaml">on: [push]
jobs:
  dispatch:
    runs-on: ubuntu-latest
    steps:
    # - run: |
    #     curl -u "${{ secrets.PAT_USERNAME}}:${{secrets.PAT_TOKEN}}" \
    #       -H "Accept: application/vnd.github.everest-preview+json" \
    #       -H "Content-Type: application/json" \
    #       https://api.github.com/repos/docker-sstc/docker-notify/actions/workflows/main.yml/dispatches \
    #       --data '{"ref": "master"}'
    - uses: actions/github-script@v6
      with:
        github-token: ${{ secrets.PAT_TOKEN }}
        script: |
          await github.rest.actions.createWorkflowDispatch({
            owner: 'docker-sstc',
            repo: 'docker-notify',
            workflow_id: 'main.yml',
            ref: 'master'
          })
</code></pre>
<p>see</p>
<ul>
<li><a href="https://octokit.github.io/rest.js/v18#actions-create-workflow-dispatch" class="inline-onebox" rel="noopener nofollow ugc">octokit/rest.js</a></li>
<li><a href="https://github.com/up9cloud/telegram-bot-send.sh/blob/master/.github/workflows/main.yml" class="inline-onebox" rel="noopener nofollow ugc">telegram-bot-send.sh/main.yml at master · up9cloud/telegram-bot-send.sh · GitHub</a></li>
</ul>

**Reply by marcchehab** [reply]
This approach using actions/github-script@v6 worked in April 2023, while the above using curl didn't work.

Thanks a lot!

**Reply by VedRatan** [reply]
the secrets.PAT_TOKEN is of the target repo or the source repo? @up9cloud

**Comment by Pixomia** [comment]
<p>Hello</p>
<p>have you found a simple Solution?</p>
<p>The Problem is simple and a default situation  nearly every developer is in.</p>
<p>Java + maven Example<br>
1 Repo with a lib  =&gt; Version 0.1.0-SNAPSHOT<br>
1 Repo with the app  =&gt;depends on repo lib =&gt; 0.1.0-SNAPSHOT<br>
both repos are own be the same developer/company</p>
<p>1 Developer commits something to the lib =&gt; i want to rebuild the app fully automated. without knowledge in the lib which app uses this lib.</p>
<p>any idea? i havent found some solution for this use case</p>

**Comment by dthomason** [comment]
<p>Appreciate you taking the time to inform us.  Definitely takes a lot of trial and error.</p>

**Comment by SkylerWittman** [comment]
I had a lot of back end forth and this is what worked for me:

We have a backend repo with a GitHub Action to pull the frontend, build node, authenticate to gcloud, and deploy to google app engine. I wanted to create an Action in my frontend that when someone merges/pushes to `development` branch we simply tell the backend to go pull the frontend `development` branch fresh and try to deploy.
The Action in the backend also runs on push to `development` branch on that side as well. That is how our CI is setup.

After a lot of back and forth this is what worked for me:

```yaml
name: Trigger Backend Action to Deploy

on:
  push:
    branches: 
      - deployment

jobs:
  send-request-to-backend:
    runs-on: ubuntu-latest
    steps:
      - name: Deploy
        run: |
          curl -X POST 'https://api.github.com/repos/{COMPANY}/{BACKEND_REPO}/actions/workflows/ci-development.yml/dispatches' \
          -H 'Accept: application/vnd.github.everest-preview+json' \
          -H 'Authorization: Bearer ${{ secrets.RUN_BACKEND_ACTION_TO_DEPLOY }}' \
          -d '{"ref":"development"}'
```

Notes: 
* The ref `development` was required as the branch on which to run this even though both scripts explicitly refer to `development` branches in both repos, but that -d was required. I didn't try it with an empty value
* I had to put _single_ quotes around the URL or else it would throw an error that it wasn't found
* `ci-development.yml` in the URL is the backend github action

**Comment by brianjmurrell** [comment]
This is all a very interesting discussion.  I am pleased that I am able to start a workflow in another repository in our org successfully with:

```yaml
      - name: Test
        uses: actions/github-script@v6
        with:
          github-token: ${{ secrets.GHA_WORKFLOW_TRIGGER }}
          script: |
            const result = await github.rest.actions.createWorkflowDispatch({
              owner: 'my_org',
              repo: 'other_repo',
              workflow_id: 'test.yml',
              ref: 'my_prs/branch'
            })
            console.log(result)
```

But per the OP, surely I want to be able to wait for the triggered workflow to complete and then use it's result as the result of the workflow that triggered it (above).  I'll go start searching on how to do that, but if anyone has an example or knows, I'd surely appreciate pointers.

Update: [trigger-workflow-and-wait](https://github.com/marketplace/actions/trigger-workflow-and-wait) seems to be working well enough.

**Reply by suryastef** [reply]
the most suitable answer so far

**Reply by brianjmurrell** [reply]
I wish it was.  I subsequently found a couple of issues with that action:

https://github.com/convictional/trigger-workflow-and-wait/issues/77
https://github.com/convictional/trigger-workflow-and-wait/issues/78

Unfortunately neither issue has received any response thus yet.  Hopefully they are fairly trivial and that action can once again be the most suitable answer (so far).  :smile:

**Reply by michaelweinold** [reply]
Is there a way to use the fantastic new `github.rest.actions.createWorkflowDispatch` with `GITHUB_TOKEN` to avoid having to re-issue a PAT every 90 days?

**Comment by jmiller-rise8** [comment]
My org has created https://github.com/marketplace/webhook-actions to allow for dispatching across repositories without the use of PATs. Currently we can only offer a free plan until we reach 100 installations.

**Comment by pbanj** [comment]
so i came up with a horrible way to do this. set up a webhook in discord to post to a private channel when the repo has the desired trigger. then have a bot watch the channel for messages and when there's a message it will trigger your repo. yes i know this is horrible

**Comment by ricekot** [comment]
It's also possible to use the GitHub CLI to send a `workflow_dispatch` event and then wait for the run to complete:
```yaml
steps:
  - run: |
      gh workflow run workflow.yml -R org/repo
      sleep 5
      gh run watch -R org/repo $(gh run list -R org/repo -w workflow.yaml -L1 --json databaseId --jq .[0].databaseId)
env:
  GH_TOKEN: ${{ secrets.MY_TOKEN }}
```
Note that you must set the `GH_TOKEN` environment variable to a token that is able to access the repository containing the workflow you're trying to run.


References: 
- https://cli.github.com/manual/gh_workflow
- https://cli.github.com/manual/gh_run

**Reply by michaelweinold** [reply]
This requires the `GH_TOKEN` to be set:

```
To use GitHub CLI in a GitHub Actions workflow, set the GH_TOKEN environment variable. Example:
  env:
    GH_TOKEN: ${{ github.token }}
```

But this token won't allow me to trigger workflows in other repositories - how are you managing to trigger those?

Are you using the "Access" settings in combination with a private repo?

![image](https://user-images.githubusercontent.com/23102087/231673084-20ff1a0b-d3b5-41ef-acd3-6940014e56ae.png)

**Reply by jmiller-rise8** [reply]
Check out [Webhook Actions GitHub App](https://github.com/marketplace/webhook-actions)

**Reply by ricekot** [reply]
Ah, thanks for pointing that out. I'm using a personal access token[^1] with the `repo` scope that's stored as a secret in the repo. I'll update my answer.
I wasn't aware of the "Access" setting but I suppose that could work too (haven't tried it though).

[^1]: https://docs.github.com/en/authentication/keeping-your-account-and-data-secure/creating-a-personal-access-token

**Reply by patrick-5546** [reply]
I found that I had to add the argument `-R org/repo` to the watch command as well

**Reply by BrianOpteran** [reply]
This is pretty much what I'd done too, but it's mental that you don't get a run ID back from the `workflow run` - GitHub is always half thought through and never complete. 🤦🏼

**Comment by OnkarRuikar** [comment]
This is how I got the repository_dispatch (not workflow_dispatch) approach mentioned in the answer to work.
Workflow in `my-org/target-repo`:
```yml
# file: target.yml

name: Gets called

on:
  repository_dispatch:
    types: [bcd_release]

jobs:
  dostuff:
    runs-on: ubuntu-latest
    steps:
      - name: "do stuff"
        env:
          VERSION: ${{ github.event.client_payload.version }}
        run: |
          echo BCD version $VERSION released!
```

Create a PAT in your user account settings. Use "Fine-grained tokens (beta)", select the `target-repo`, give read-write access to "Content" category, copy the token code and put it in a secret(DISPATCH_PAT) in `my-org/source-repo`.

Workflow in `my-org/source-repo`:
```yml
#file: source.yml

name: Release BCD

# I am triggering this manually so using 'workflow_dispatch'
on:
  workflow_dispatch

jobs:
  release:
    runs-on: ubuntu-latest
    steps:
    - uses: actions/github-script@v6
      with:
        github-token: ${{ secrets.DISPATCH_PAT }}
        script: |
          const result = await github.rest.repos.createDispatchEvent({
            owner: 'my-org',
            repo: 'target-repo',
            event_type: 'bcd_release',
            client_payload: {"version": "v5.3.10"}
          })
          console.log(result);
```
---

Here is our working live implementation:
- Source repo, `mdn/browser-compat-data`, makes a [repository_dispatch request named `bcd_release`](https://github.com/mdn/browser-compat-data/blob/34217bc575e78b157b0042c8b842795b66cda86a/.github/workflows/update_bcd-utils_api.yml#L18) after the package is published to NPM.
- In target repo, `mdn/bcd-utils`, multiple workflows gets triggered by the `bcd_release` event. For example, [Prod Build - API ](https://github.com/mdn/bcd-utils/blob/a746468d5d432f4f6b1190156ecf4e750989f5e8/.github/workflows/deploy-prod-api.yml#L16)  and [Prod Build - Updates](https://github.com/mdn/bcd-utils/blob/a746468d5d432f4f6b1190156ecf4e750989f5e8/.github/workflows/deploy-prod-updates.yml#L16) workflows.

**Comment by Sithu99-dev** [comment]
How can I apply this to a pair of repositories within an organization?

**Comment by tbish967** [comment]
I'm not too hopeful, but does anyone know if there's any update on a standard way for a consumer repo to monitor for a version update on a producer repo, _without_ the producer taking some action (creating a dispatch event)? My main issues with this are:

* New consumers need to be registered with the producer on creation somehow; likely on some external artifact accessible by both workflows
* This workflow isn't applicable if the producer has no knowledge of its consumers, which should be the case

The closest I've found is a cron triggered workflow running nightly to poll that producer for a new version; if I want versions faster, it seems like this requires some custom infrastructure.

**Reply by OnkarRuikar** [reply]
Just a thought. If your consumer wants to monitor a package on NPM then, one way is to use [npmjs.com's hooks](https://docs.npmjs.com/cli/v7/commands/npm-hook).

Related packages:
- https://www.npmjs.com/package/hooks
- https://www.npmjs.com/package/npm-hook-receiver
  The receiver can then make a POST request sending `repository_dispatch` event to consumer.

For receiver you'll need a self-hosted runner or a dedicated instance(which will send `repository_dispatch` event to consumer repo).

**Comment by karthickradhakrishnan** [comment]
Along with the trigger is it possible to pass a value from repo to another?? Eg: Pass docker image tag to be passed from one to another repo

**Reply by ARolek** [reply]
I'm investigating this same idea. I would like one repo to focus on building docker images and publishing them to a registry, and another repo to be responsible for deployment (via Terraform). Being able to trigger a workflow in another repo and pass context would allow for this.

**Reply by ARolek** [reply]
@karthickradhakrishnan following up this, it's possible to pass params to the action. Take a look at the [docs for gh workflow run](https://cli.github.com/manual/gh_workflow_run):

> If the workflow file supports inputs, they can be specified in a few ways:
> 
> * Interactively
> * via -f or -F flags
> * As JSON, via STDIN

**Reply by karthickradhakrishnan** [reply]
Thanks @ARolek will check on this.

**Comment by pedrolamas** [comment]
Not exactly a solution to the requested, but there is always the option to schedule the running of a workflow in your Application repo to check for changes in the Library repo, and when it finds changes, perform a build!

I use this approach on a couple of repositories of mine, specifically:

- https://github.com/pedrolamas/docker-mainsail/blob/5d355784876de54a75af5733c80545c08898b1b1/.github/workflows/main.yml#L5
- https://github.com/pedrolamas/docker-klipper-simulavr/blob/348be296310f4e83d1f7de804a79c89afe4c49d6/.github/workflows/main.yml#L5

In both cases, I check the latest commit to the referenced repo, calculate a hash of it, and check if I have a docker container with that hash; if I do, the workflow ends, but if such image does not exist, it will perform the build action to the end!

**Reply by ericreg-butlr** [reply]
Thank you. This is the right answer. With your solution, the entire pipeline doesn't break when someone accidentally breaks the `workflow_dispatch` in the produce repository.

**Comment by datsculptor** [comment]
You can set up GitHub Actions to run CI in both repositories upon a push to one. Just use repository dispatch events to trigger workflows in the other repository.

**Comment by zauguin** [comment]
If someone else encounters this, when I had this issue I started writing the [zauguin/trigger-action-event](https://github.com/zauguin/trigger-action-event) action. This action can be used in some upstream repository to trigger events, then other repositories can use a [related GitHub App](https://github.com/apps/cross-repo-action-triggers) to subscribe to these events and get actions triggered on them. That way no tokens need to be exchanged manually and the repository triggering the change does not need to be aware of all consumers.
